src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -109,14 +117,124 @@ protected Object readResolve() throws ObjectStreamException {
+     * For old {@code workflow-job}, the top-level entries will be missing;
+     * for old {@code workflow-cps}, the second-level entries will be missing.
+     */
+    @Restricted(DoNotUse.class)
+    @Extension public static final class GraphL implements GraphListener.Synchronous {
+        @Override public void onNewHead(FlowNode node) {
+            LOGGER.finer("GraphListener working");
+            if (node instanceof BlockStartNode || node instanceof BlockEndNode) {
+                Map<String, Boolean> startNodesAreClosed = FlowL.startNodesAreClosedByFlow().get(node.getExecution().getOwner());
+                if (startNodesAreClosed != null) {
+                    if (node instanceof BlockStartNode) {
+                        assert !startNodesAreClosed.containsKey(node.getId());
+                        startNodesAreClosed.put(node.getId(), false);
+                    } else {
+                        startNodesAreClosed.put(((BlockEndNode) node).getStartNode().getId(), true);
+                    }
+                }
+            }
+        }
+    }
+    @Restricted(DoNotUse.class)
+    @Extension public static final class FlowL extends FlowExecutionListener {
+        final Map<FlowExecutionOwner, Map<String, Boolean>> startNodesAreClosedByFlow = new HashMap<>();
+        static Map<FlowExecutionOwner, Map<String, Boolean>> startNodesAreClosedByFlow() {
+            FlowL flowL = ExtensionList.lookup(FlowExecutionListener.class).get(FlowL.class);
+            return flowL != null ? flowL.startNodesAreClosedByFlow : /* ? */ new HashMap<FlowExecutionOwner, Map<String, Boolean>>();
+        }
+        @Override public void onRunning(FlowExecution execution) {
+            LOGGER.finer("FlowExecutionListener working");
+            assert !startNodesAreClosedByFlow.containsKey(execution.getOwner());
+            startNodesAreClosedByFlow.put(execution.getOwner(), new HashMap<String, Boolean>());
+        }
+        @Override public void onResumed(FlowExecution execution) {
+            assert !startNodesAreClosedByFlow.containsKey(execution.getOwner());
+            Map<String, Boolean> startNodesAreClosed = new HashMap<String, Boolean>();
+            startNodesAreClosedByFlow.put(execution.getOwner(), startNodesAreClosed);
+            DepthFirstScanner dfs = new DepthFirstScanner();
+            dfs.setup(execution.getCurrentHeads());
+            for (FlowNode n : dfs) {
+                if (n instanceof BlockEndNode) {
+                    startNodesAreClosed.put(((BlockEndNode) n).getStartNode().getId(), true);
+                } else if (n instanceof BlockStartNode) {
+                    if (!startNodesAreClosed.containsKey(n.getId())) {
+                        startNodesAreClosed.put(n.getId(), false);
+                    }
+                }
+            }
+        }
+        @Override public void onCompleted(FlowExecution execution) {
+            assert startNodesAreClosedByFlow.containsKey(execution.getOwner());
```

Comment:
> we're marking all nodes closed when we see the end, but for start nodes only toggle them if not present

I guess you are referring to `onResumed` here, not `onCompleted` where the comment was placed?